### PR TITLE
Fix missing bytesize check for safe `find` vectorization with LLVM

### LINKED ISSUE #6294: `<xutility>`: std::find no longer compiles with structs with defaulted equality comparison with Clang-CL
# Describe the bug
`std::find` and `std::ranges::find` (both standardized in `<algorithm>` but implemented in `<xutility>`) have long been able to handle structs with defaulted equality comparison (behavior extended in the C++20 standard). However, recent optimizations regarding LLVM usage have resulted in a regression (https://github.com/microsoft/STL/issues/5479 and specifically https://github.com/microsoft/STL/pull/5767/changes#diff-980098f27df5ad089b2a93a8489574cfcab50f4b9d9c55524b0a57444a4c664eR6137-R6138). In particular, types with a `sizeof` other than 1, 2, 4 or 8, that have a defaulted equality comparison operator, fail to compile.

This was checked using the latest version of the STL shipped with Visual Studio `18.6.0`, which seems to be `19.51.36231`, and Clang-CL version `20.1.8`.

# Command-line test case

```
C:\Temp>type repro.cpp
#include <algorithm>
#include <cstdint>
#include <vector>

namespace {
struct A {
  int64_t a{};
  int64_t b{};

  friend constexpr bool operator==(A x, A y) = default;
};
struct B {
  int64_t a{};
  int64_t b{};

  friend constexpr bool operator==(B x, B y) {
    return x.a == y.a && x.b == y.b;
  }
};
} // namespace

int main() {
  // the following calls fail to compile due to defaulted comparison operator
  const auto a = std::vector<A>{};
  std::ignore = std::find(a.begin(), a.end(), A{}); // does not compile
  std::ignore = std::ranges::find(a, A{});          // does not compile

  // user-defined comparison operator works fine
  const auto b = std::vector<B>{};
  std::ignore = std::find(b.begin(), b.end(), B{}); // OK
  std::ignore = std::ranges::find(b, B{});          // OK

  return 0;
}

C:\Temp>clang-cl -clang:-std=c++23 -- repro.cpp 
In file included from repro.cpp:1:
In file included from C:\Program Files\Microsoft Visual Studio\18\Enterprise\VC\Tools\MSVC\14.51.36231\include\algorithm:10:
In file included from C:\Program Files\Microsoft Visual Studio\18\Enterprise\VC\Tools\MSVC\14.51.36231\include\__msvc_heap_algorithms.hpp:11:
C:\Program Files\Microsoft Visual Studio\18\Enterprise\VC\Tools\MSVC\14.51.36231\include\xutility(320,23): error: static assertion failed: unexpected size
  320 |         static_assert(false, "unexpected size");
      |                       ^~~~~
C:\Program Files\Microsoft Visual Studio\18\Enterprise\VC\Tools\MSVC\14.51.36231\include\xutility(6445,42): note: in instantiation of function template specialization 'std::_Find_vectorized<const (anonymous namespace)::A,
      (anonymous namespace)::A>' requested here
 6445 |             const auto _Result    = _STD _Find_vectorized(_First_ptr, _STD _To_address(_Last), _Val);
      |                                          ^
1 error generated
```

NOTE: the above script compiles just fine with MSVC shipped with the same version of Visual Studio

```
C:\Temp>cl /EHsc /W4 /WX /std:c++latest .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.51.36243 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

/std:c++latest is provided as a preview of language features from the latest C++
working draft, and we're eager to hear about bugs and suggestions for improvements.
However, note that these features are provided as-is without support, and subject
to changes or removal as the working draft evolves. See
https://go.microsoft.com/fwlink/?linkid=2045807 for details.

repro.cpp
Microsoft (R) Incremental Linker Version 14.51.36243.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:repro.exe 
repro.obj 
```

I found that the vectorization is only implemented on 1-, 2-, 4- or 8-bytes objects, and that all other sizes trigger the above `error: static assertion failed: unexpected size`.

However, because `_Vector_alg_in_find_is_safe_elem` under the hood uses `__is_trivially_equality_comparable`, which evaluates to `true`, it takes the vectorized branch anyways, which, in turn, triggers the static assertion.

# Expected behavior
`std::find` and `std::ranges::find` work on types with comparison operator for both Clang-CL and MSVC, regardless of whether it was defaulted or explicitly specified.

Either:

* [ ] the vectorization needs to be implemented for types with a size other than 1, 2, 4 or 8 bytes; or
* [ ] `_Vector_alg_in_find_is_safe` needs to explicitly exclude those odd-sized and/or large types.

# STL version
* Option 2: `_MSVC_STL_UPDATE` value (because Clang-CL is used)
  + Example:
    ```
    C:\Temp>type meow.cpp
    #include <iostream>
    int main() {
        std::cout << _MSVC_STL_UPDATE << "\n";
    }

    PS C:\Temp> clang-cl -clang:-std=c++23 -MDd -c -- .\meow.cpp
    PS C:\Temp> lld-link.exe meow.obj /out:meow.exe      
    PS C:\Temp> .\meow.exe                                      
    202604
    ```

* For completeness: `cl.exe` shows `Microsoft (R) C/C++ Optimizing Compiler Version 19.51.36243 for x64`
  + Example:
    ```
    C:\Temp>type meow.cpp
    #include <iostream>
    int main() {
        std::cout << _MSVC_STL_UPDATE << "\n";
    }
    C:\Temp>cl /EHsc /nologo /W4 meow.cpp && meow
    meow.cpp
    202604
    ```

# Additional context
* This was found in our Nightly build on GitHub (https://github.com/PowerGridModel/power-grid-model/actions/runs/26138748913/attempts/1) which runs on the `msvc-latest` GitHub Actions runner image. Now that the latest Visual Studio is being deployed in the GitHub Actions runner images (https://github.com/actions/runner-images/blob/win25-vs2026/20260518.113/images/windows/Windows2025-VS2026-Readme.md), this will affect more users.
* https://en.cppreference.com/cpp/language/default_comparisons contains an example with a struct similar to what was used here, so that suggests that it really is something that is supposed to be supported.
* `std::pair<int, int>` does not suffer from this issue (contrary to what is suggested in https://github.com/microsoft/STL/issues/54790) because it does not have a defaulted equality comparison operator:
  ```cpp
  _EXPORT_STD template <class _Ty1, class _Ty2, class _Uty1, class _Uty2>
  _NODISCARD constexpr bool operator==(const pair<_Ty1, _Ty2>& _Left, const pair<_Uty1, _Uty2>& _Right) {
     return _Left.first == _Right.first && _Left.second == _Right.second;
  }
  ```
  This also implies that `std::pair<int, int>` at the moment is not vectorized.
* If I understand the libcxx implementation of `find` for types satisfying `__is_trivially_equality_comparable_v` (https://github.com/llvm/llvm-project/blob/main/libcxx/include/__algorithm/find.h#L137-L139) correctly, then it looks like LLVM opts for the exclusion of vectorization for types with sizes other than `1` or `sizeof(wchar)` and that also are not integral types (which effectively excludes more than aforementioned approach `_Vector_alg_in_find_is_safe needs to explicitly exclude those odd-sized and/or large types.`)

### PR DESCRIPTION
**Pull request by mgovers**
Fixes #6294 

Here's my attempt at excluding odd-sized and large-sized structs from the vectorization in `xutility`.

## Design approach

I have tried to follow all the conventions used elsewhere as much as possible, in particular:

* for the implementation, I followed what was already there and the changes from #5767
* for the tests, following https://github.com/microsoft/STL/commit/3da7d27db7d714baae5eb5ce88603aa26fb44ccd
  * NOTE: I explicitly scoped out equivalent tests as for `_Equal_memcmp_is_safe` as the current `main` implementation of `_Vector_alg_in_find_is_safe` was not tested in the first place. I still opted to add a couple more tests than for `_Vector_alg_in_search_is_safe` to at the very least show that there's vectorization for MSVC and not only for Clang
* Since `_Vector_alg_in_search_is_safe` already had a similar check as what was needed here, I re-used that one (had to move it outside the `#ifdef` statement, though)
  * I did multiple searches for strings like `_VECTORIZED_FIND`, `static_assert(false, "unexpected size");`, `_vectorized(` and `_Nx <= 8 && (_Nx & (_Nx - 1)) == 0` to see if there were other cases that I missed. I believe (and hope) I got them all (at least in `algorithm` and `xutility`).
  * I intentionally did not include `MINMAX`-like things, because they are independently explicitly guarded where the vectorized versions are used.

## Additional remarks/open questions

* I am re-running all tests locally once more as we speak, but it seems like some unrelated tests are failing on my device, also on `main`, so those are not an issue with my branch.
  * I will compare the failed scenarios once I'm finished running.
  * I am already opening this PR in the hope that CI will tell me for sure.
* What is the best way (if any) to add the test case from #6294 ?
  * Is it worth the effort to create a very small test file that contains exactly that edge case?
  * Or should I add it to the one I already added?
* Should I benchmark? Or will that be done on your end?
  * (cfr. `For performance-sensitive code (e.g. containers and algorithms) you may wish to write and/or run benchmarks, and the STL team will likely run any benchmarks we do have in our PR process.`)
  * It is touching sensitive code (very common algorithms with high performance demands) and explicitly disabling vectorization (which is, in fact, a bugfix and it **should** only be excluding things that never compiled in the first place, but still...)

Despite my questions, I think (and hope) that my contribution is in a good state anyways.

### CODE REVIEW

File: tests/std/tests/GH_006294_vectorized_find_is_safe/test.compile.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
+
+#if _HAS_CXX20
+
+#ifdef __clang__
+constexpr bool vectorized_trivial_comparison = true;
+#else
+constexpr bool vectorized_trivial_comparison = false;
+#endif
```
[the remark this answers is not in the document]
Comment (mgovers) [reply tests/std/tests/GH_006294_vectorized_find_is_safe/test.compile.pass.cpp:17]: I don't think there's such a trait. Instead, since the actual implementation uses `#ifdef __clang__`, I believe that this approach is the correct one. It also aligns with #5767

**Comment by mgovers** [comment]
The failing tests are x86_86-only:

```
Failed Tests (14):
  libc++ :: std/containers/container.adaptors/flat.map/flat.map.access/index_transparent.pass.cpp:0
  libc++ :: std/containers/container.adaptors/flat.set/flat.set.capacity/size.pass.cpp:2
  libc++ :: std/containers/container.adaptors/queue/queue.cons/move_noexcept.pass.cpp:2
  libc++ :: std/time/time.cal/time.cal.mwdlast/types.pass.cpp:2
  libc++ :: std/time/weeks.pass.cpp:0
  std :: tests/Dev11_0732166_unordered_strong_guarantee:41
  std :: tests/P0009R18_mdspan_layout_right:21
  std :: tests/P1206R7_unordered_multiset_from_range:27
  std :: tests/P2286R8_text_formatting_range_formatter:21
  std :: tests/P2321R2_views_zip_transform:027
  std :: tests/P2321R2_views_zip_transform:077
  std :: tests/P3349R1_contiguous_iterators_to_pointers:28
  std :: tests/VSO_0000000_vector_algorithms:037
  std :: tests/VSO_0961751_hash_range_erase:25
```

However, I do not have an x86 machine. How can I investigate what went wrong?

Regarding the CLA, I did not forget about it, I'm on it.

**Comment by AlexGuteniev** [comment]
> However, I do not have an x86 machine. How can I investigate what went wrong?

x64 OS can still execute x86 program using [Wow64](https://learn.microsoft.com/en-us/windows/win32/winprog64/wow64-implementation-details). You don't need to do anything special for that, just run x86 executables.

Specifically, you just need to follow x86 instructions from [README.md](https://github.com/microsoft/STL#how-to-build-with-the-command-prompt) on x64 machine.

**Comment by mgovers** [comment]
> > However, I do not have an x86 machine. How can I investigate what went wrong?
> 
> x64 OS can still execute x86 program using [Wow64](https://learn.microsoft.com/en-us/windows/win32/winprog64/wow64-implementation-details). You don't need to do anything special for that, just run x86 executables.
> 
> Specifically, you just need to follow x86 instructions from [README.md](https://github.com/microsoft/STL#how-to-build-with-the-command-prompt) on x64 machine.

I am now able to run the x86 tests but don't seem to be able to reproduce the error...

from an x86 native tools command prompt, running for all failing tests in the above:

```ps
<...>\STL\out\x86> python tests\utils\stl-lit\stl-lit.py -Dnotags=ASAN ..\..\tests\<test_dir>\
```

yields

```txt
PASS: std :: tests/Dev11_0732166_unordered_strong_guarantee:41 (32 of 46)
PASS: std :: tests/P0009R18_mdspan_layout_right:21 (27 of 30)
PASS: std :: tests/P1206R7_unordered_multiset_from_range:27 (28 of 30)
PASS: std :: tests/P2286R8_text_formatting_range_formatter:21 (26 of 30)
PASS: std :: tests/P2321R2_views_zip_transform:027 (52 of 116)
PASS: std :: tests/P2321R2_views_zip_transform:077 (98 of 116)
PASS: std :: tests/P3349R1_contiguous_iterators_to_pointers:28 (62 of 64)
PASS: std :: tests/VSO_0000000_vector_algorithms:037 (112 of 138)
PASS: std :: tests/VSO_0961751_hash_range_erase:25 (37 of 46)
```

I am still running the libc++ tests (for reference; will update once i have the results):

```
PASS: libc++ :: std/containers/container.adaptors/flat.map/flat.map.access/index_transparent.pass.cpp:0 (2444 of 24345)
PASS: libc++ :: std/containers/container.adaptors/flat.set/flat.set.capacity/size.pass.cpp:2 (3239 of 24345)
PASS: libc++ :: std/containers/container.adaptors/queue/queue.cons/move_noexcept.pass.cpp:2 (3534 of 24345)
PASS: libc++ :: std/time/time.cal/time.cal.mwdlast/types.pass.cpp:2 (19371 of 24345)
PASS: libc++ :: std/time/weeks.pass.cpp:0 (20511 of 24345)
```

**Comment by AlexGuteniev** [comment]
> from an x86 native tools command prompt, running for all failing tests in the above:

Then I suspect it is a spurious CI failure.

Observe that only one shard has failed tests, whereas I'd expect them to be distributed across 10 of them.

Let's assume it passed, and you only owe CLA here.

**Comment by mgovers** [comment]
> > from an x86 native tools command prompt, running for all failing tests in the above:
> 
> Then I suspect it is a spurious CI failure.
> 
> Observe that only one shard has failed tests, whereas I'd expect them to be distributed across 10 of them.
> 
> Let's assume it passed, and you only owe CLA here.

OK, awesome, thank you!

Yes, I am working with the OSPO to get the paperwork for the CLA finished at my company internally. We expect that process to be finished tomorrow.

After that, I will start the process here and make sure the appropriate people take the right follow-up steps to finish it (should not take long).

Do we need to run the benchmarks as well?

**Comment by AlexGuteniev** [comment]
> Do we need to run the benchmarks as well?

I assume that you don't need to write your own benchmark, but it would be good to run existing benchmarks before and after your changes, for `find`, just to make sure the vectorization code path is still enabled after your changes.

Observe that out of all #5479 PRs I only added a new benchmark in #5591, to see if adding odd size support really helps

I'm not a maintainer though, I only contributed a part of the vectorization here.

**Comment by StephanTLavavej** [comment]
I reran the failed shard and it passed; it looked like the machine got corrupted and it wasn't our test suite's fault (much less yours).

I am not terribly concerned about running benchmarks for a fix like this; the benchmarks are useful when making tuning decisions.

I am severely overloaded at the moment (half of my time is going to non-STL work - not my choice, sorry!) but your PR is on my radar and I'll review it when I get a chance. The only feedback I have right now is to remove the bug number from the PR title, where it provides no value. (The PR description saying `Fixes #nnnn` is perfect.)

**Comment by mgovers** [comment]
> I reran the failed shard and it passed; it looked like the machine got corrupted and it wasn't our test suite's fault (much less yours).
>
> I am not terribly concerned about running benchmarks for a fix like this; the benchmarks are useful when making tuning decisions.

Great! Thanks for the input

> I am severely overloaded at the moment (half of my time is going to non-STL work - not my choice, sorry!) but your PR is on my radar and I'll review it when I get a chance. 

No worries, same for me and the project I'm working on

> The only feedback I have right now is to remove the bug number from the PR title, where it provides no value. (The PR description saying `Fixes #nnnn` is perfect.)

Done ✅

**Comment by mgovers** [comment]
@microsoft-github-policy-service agree company="Alliander"

**Comment by StephanTLavavej** [comment]
Apologies for the delay in reviewing here, I'm digging out from a backlog that accumulated while I was tasked with non-STL work for a couple of months, and I need to get `constexpr <cmath>` finished first.

**Comment by mgovers** [comment]
No worries! It happens, we've all been there. `constexpr` `<cmath>` is worth the wait!
